File: compiler/bench/stack.c
// expected: 1500000000
#include <stdio.h>
int main() {
    long a = 0, b = 1, c = 2, t;
    for (long i = 0; i < 500000000; i++) {
        // rot: a b c -> b c a
        t = a; a = b; b = c; c = t;
        // + swap -: b c a -> b c+a -> c+a b -> c+a-b b
        c = c + a;
        t = b; b = c; c = t;
        a = c - b;
        // over +: restore balance
        a = a + b;
    }
    printf("%ld\n", a + b + c);
    return 0;
}

File: compiler/bench/stack.fs
\ expected: 1500000000
\ Stack manipulation stress - rot, -rot, 2swap, nip, tuck

: main
  0 1 2 500000000 0 do
    rot + swap -     \ a b c -> b c a -> b c+a -> c+a b -> c+a-b b
    over +           \ restore balance
  loop
  + + . cr ;
